### PR TITLE
[CPU][x86] add support for sink input of SDPA

### PR DESCRIPTION
**Pull request by tiger100256-hu**
### Details:
 - *extend the softmax kernel API to make it support sink input*
 - *in the reference implementation of sink input, there are 3 operations, **cat,** **softmax**, **drop**, because **cat** and **drop** will make attention_weight shape changed,   so modify softmax API maybe a good way to reduce memory allocation and movement*
 - reference1: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt_oss/modeling_gpt_oss.py#L258-L264
 - reference2: https://github.com/openvinotoolkit/openvino/pull/31786/files

### Tickets:
 - *CVS-173533*

### CODE REVIEW

File: src/plugins/intel_cpu/src/utils/plain_tensor.hpp [hunk 1]
```diff
@@ -439,6 +439,35 @@ struct PlainTensor {
         return reinterpret_cast<DT*>(m_ptr.get()) + off;
     }
 
+    template <int dim>
+    [[nodiscard]] int64_t safe_offset() const {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/utils/plain_tensor.hpp]: Why add those safe_offset ? You could check if `PlainTensor` is empty outside

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp [hunk 2]
```diff
@@ -853,6 +868,9 @@ struct MHAKernel<ScaledDotProductAttention::KT_MLAS, float> {
             for (size_t m = m_start; m < m_end; m++) {
                 // apply attention mask & sofmax
                 auto ncausal = auto_causal ? (kv_len - q_len + m + 1) : kv_len;
+
+                void* sink = reinterpret_cast<void*>(sink_input.safe_ptr<float>(b, h, m, 0));
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/scaled_attn.cpp]: If sink is always float type, why do we need to cast it to void ? If `sink_input` is the same type with `q`, you could use the same template type of Q to cast the `sink_input`.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.cpp [hunk 3]
```diff
@@ -1591,6 +1592,8 @@ static void mha_single_token_kernel(const ov::intel_cpu::PlainTensor& query,
             attn_mask_ptr = reinterpret_cast<uint8_t*>(&attention_mask.at<T>({b, h, pq, 0}, true));
         }
         uint8_t* cmask_ptr = causal_mask ? &causal_mask.at<uint8_t>({b, h, pq, 0}, true) : nullptr;
+
+        auto sink = sink_input.safe_ptr<T3>(b, h, pq);
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.cpp]: I think the `safe_ptr` way doesn't align with usage of `PlainTensor`. You could just check if `sink_input` is a `nullptr`, `Boolean Operator` is overridden in PlainTensor`
```suggestion
        auto sink = sink_input.safe_ptr<T3>(b, h, pq);
        auto sink = sink_input ? sink_input.safe_ptr<T3>(b, h, pq) : nullptr;
```

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp [hunk 4]
```diff
@@ -1068,7 +1090,8 @@ inline void attn_softmax_kernel(T* a,
                                 size_t total_size,
                                 ov::element::Type attn_mask_prec,
                                 ov::element::Type dst_precision,
-                                float alibi_slope = 0);
+                                float alibi_slope = 0,
+                                T* sink = nullptr);
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: Will sink be other types rather than `float` ? It seems  we don't write kernels to support other data types such as `float16/bfloat16`

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp [hunk 5]
```diff
@@ -475,7 +481,8 @@ inline void scale_add2_reduce_max(ov::float16* a,
                                   bool select_nfltmax_at_0,  // true:  0 in mask set -FLT_MAX
                                   size_t size,
                                   float alibi_slope,
-                                  ov::float16& max) {
+                                  ov::float16& max,
+                                  float* sink = nullptr) {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: I think it's better not to add default value to this template function, you should follow the style of `has_alibi`, `has_attn_mask` and `has_causal_mask`, you can add a template arg called `has_sink`, and only check this template arg rather than sink. Template check is executed during compile time, it has better performance.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp [hunk 6]
```diff
@@ -696,6 +703,13 @@ inline void exp_reduce_sum(float* a, const float max, const size_t size, float&
 
         i += (size - i);
     }
+    if (sink != nullptr) {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: If add a template arg `has_sink` like others, the check here could be changed.
```suggestion
    if (sink != nullptr) {
    if constexpr (has_sink) {
```

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp [hunk 7]
```diff
@@ -721,6 +735,14 @@ inline void exp_reduce_sum(float* a, const float max, const size_t size, float&
...
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: same with above

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp [hunk 8]
```diff
@@ -1185,7 +1212,8 @@ inline void attn_softmax_kernel<ov::float16>(ov::float16* a,
                                              size_t total_size,
                                              ov::element::Type attn_mask_prec,
                                              ov::element::Type dst_precision,
-                                             float alibi_slope) {
+                                             float alibi_slope,
+                                             ov::float16* sink = nullptr) {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: it's better to pass sink to the inner function, otherwise outside users may find that sink parameter doesn't take effect

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp [hunk 9]
```diff
@@ -191,7 +191,8 @@ inline void scale_add2_reduce_max(float* a,
                                   bool select_nfltmax_at_0,  // true:  0 in mask set -FLT_MAX
                                   size_t size,
                                   float alibi_slope,
-                                  float& max) {
+                                  float& max,
+                                  float* sink = nullptr) {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: consider not to add default value, let outside user to carefully pass correct value of sink.

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp [hunk 10]
```diff
@@ -122,13 +122,19 @@ struct MHAKernel {
         return result;
     }
 
-    void softmax(float* a, int len) {
+    void softmax(float* a, int len, float* sink = nullptr) {
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/scaled_attn.cpp]: remove default value, let outside user decides the right value of sink.

File: src/plugins/intel_cpu/src/utils/plain_tensor.hpp (hunk 1 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/utils/plain_tensor.hpp]: the sink input shape is not totally match with attention_weight, for example the attention_weight shape is (1, 64, 10, 10), and the sink shape is (1, 64, 1, 1), so I need to calculate the index of sink input carefully and make it not out of range. if the index of attention_weight is b=0, h=63, m=9, so need to check if b, h , m are out of range of sink_input, there are multi places to call sink_input.ptr(b, h, m), so I add **safe_ptr** to simple these checking

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp (hunk 2 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp]: fixed, only use float*

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp (hunk 4 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: fixed, only use float*

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp (hunk 5 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: fixed, there is only one sink input, so use c++ code to do it

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp (hunk 6 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: fixed, there is only one sink input, so use c++ code to do it

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp (hunk 9 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp]: fixed

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp (hunk 10 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp]: fixed

File: src/plugins/intel_cpu/src/utils/plain_tensor.hpp (hunk 1 above)
Comment (zhangYiIntel) [reply src/plugins/intel_cpu/src/utils/plain_tensor.hpp]: Maybe we could consider use `%` to find the real index, `minus - 1` may not always be correct

Comment (zhangYiIntel) [reply src/plugins/intel_cpu/src/utils/plain_tensor.hpp]: After analysis the shape of sink input, I find that it has layout `[Batch, Head_Num, Seq_length, Head_Size]`, `head_num` and `head_size` is fixed for certain LLM, but `batch/seq_length` is dynamic during inference. So I think we should check if `batch/seq_length` is numpy-broadcastable.  So for `batch/seq_length` we should check if it is `1`, if so, it means that dim should use broadcast, otherwise real index is used. For example,
```
Q [4, 64, 128, 64]
// no-broadcastbale in sink
Sink [4, 64, 128, 64]
// broadcast in batch/seq_length dims.
Sink [1, 64, 1, 64]
```
After applying this check, there is also no need to change `PlainTensor`

File: src/plugins/intel_cpu/src/shape_inference/custom/scaled_attn.cpp [hunk 11]
```diff
@@ -88,6 +88,55 @@ class SDPAShapeInfer : public ShapeInferEmptyPads {
...
+            auto weight_dims_size = weight_dims.size();
+            if (weight_dims_size != sink_dims.size()) {
+                sink_ok = false;
+            } else {
+                weight_dims[3] = present_v_dims[length_index];
+                auto check_broadcast = [](const size_t& target, const size_t& to) -> bool {
+                    return any_of(target, to, 1U);
+                };
+                for (size_t i = 0; i < weight_dims_size - 2; i++) {
+                    sink_ok = sink_ok && check_broadcast(sink_dims[i], weight_dims[i]);
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/shape_inference/custom/scaled_attn.cpp]: The sink input has shape `[batch, head_num, seq_len, 1]`
Only `Batch`/`Seq_len` is broadcastable, I think we could just check whether those 2 dims is broadcastable. And head_num should be the same with `query`.
if the input_layout is `[batch, head_num, seq_length, head_size]`, the `weight_dims_size - 2` here only checks the `batch` and 'head_num'.

Comment (zhangYiIntel) [reply src/plugins/intel_cpu/src/shape_inference/custom/scaled_attn.cpp]: Also there is a limitation in the `check_broadcast`. The following shape is also valid but the check here will throw exception. Broadcast here means that any_of them is 1 or they have same value.
```
Query [1, 64, 3, 1]
Sink [1, 64, 3 ,1]
```
For checking the broadcast of sinks I think this criteria is better
```
auto q_len = query_shape[2];
auto sink_len = sink_shape[2];
bool is_len_broadcast = q_len == sink_len | sink_len ==1;
```

**Comment by yuxu42** [comment]
Hi @maxnick could you please take a review? Thanks!

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.hpp [hunk 12]
```diff
@@ -26,6 +26,7 @@ void mha_single_token(const ov::intel_cpu::PlainTensor& query,
                       ov::intel_cpu::PlainTensor& head_sum,
                       size_t key_group_size,
                       size_t value_group_size,
-                      bool quant_key_by_channel);
+                      bool quant_key_by_channel,
+                      const ov::intel_cpu::PlainTensor& sink_input);
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.hpp:38]: Please add a short comment about the purpose of this sink input and what should be set if there is no sink input.

**Comment by maxnick** [comment]
PR is waiting performance checks.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.hpp (hunk 12 above)
Comment (tiger100256-hu) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.hpp:38]: already add comment, please help to check it, thanks
